# HG changeset patch
# User John W. Eaton <jwe@octave.org>
# Date 1387845274 18000
#      Mon Dec 23 19:34:34 2013 -0500
# Branch stable
# Node ID 65e4e0569ed4ed2aa368f0048561f0b9c19d1f81
# Parent  bce3a82a4a8de81b81a0d0f26dd8a837932b2bbb
install octave-gui binary in libexec directory

* src/Makefile.am (archlib_PROGRAMS): New variable.  List octave-gui
* here, not in OCTAVE_BINARIES (and, ultimately, bin_PROGRAMS).
* src/main.in.cc (OCTAVE_ARCHLIBDIR): New macro.
(get_octave_archlibdir): New function.
(main): New variable octave_archlibdir.  If HAVE_OCTAVE_GUI, set file
to octave-gui in the octave_archlibdir directory.

diff --git a/src/Makefile.am b/src/Makefile.am
--- a/src/Makefile.am
+++ b/src/Makefile.am
@@ -47,17 +47,17 @@ DISTCLEANFILES = \
 
 OCTAVE_BINARIES = \
   octave \
   octave-cli
 
 OCTAVE_VERSION_LINKS = octave-cli-$(version)$(EXEEXT)
 
 if AMCOND_BUILD_GUI
-  OCTAVE_BINARIES += octave-gui
+  archlib_PROGRAMS = octave-gui
   OCTAVE_VERSION_LINKS += octave-gui-$(version)$(EXEEXT)
 endif
 
 OCTAVE_CORE_LIBS = \
   $(top_builddir)/libinterp/liboctinterp.la \
   $(top_builddir)/liboctave/liboctave.la
 
 include ../libgui/link-deps.mk
diff --git a/src/main.in.cc b/src/main.in.cc
--- a/src/main.in.cc
+++ b/src/main.in.cc
@@ -41,16 +41,20 @@ along with Octave; see the file COPYING.
 
 #include <sys/types.h>
 #include <unistd.h>
 
 #ifndef OCTAVE_VERSION
 #define OCTAVE_VERSION %OCTAVE_VERSION%
 #endif
 
+#ifndef OCTAVE_ARCHLIBDIR
+#define OCTAVE_ARCHLIBDIR %OCTAVE_ARCHLIBDIR%
+#endif
+
 #ifndef OCTAVE_BINDIR
 #define OCTAVE_BINDIR %OCTAVE_BINDIR%
 #endif
 
 #ifndef OCTAVE_PREFIX
 #define OCTAVE_PREFIX %OCTAVE_PREFIX%
 #endif
 
@@ -434,16 +438,30 @@ get_octave_bindir (void)
   // has been relocated to some installation directory other than the
   // one originally configured.
 
   std::string obd = octave_getenv ("OCTAVE_BINDIR");
 
   return obd.empty () ? subst_octave_home (std::string (OCTAVE_BINDIR)) : obd;
 }
 
+static std::string
+get_octave_archlibdir (void)
+{
+  // Accept value from the environment literally, but substitute
+  // OCTAVE_HOME in the configuration value OCTAVE_ARCHLIBDIR in case
+  // Octave has been relocated to some installation directory other than
+  // the one originally configured.
+
+  std::string dir = octave_getenv ("OCTAVE_ARCHLIBDIR");
+
+  return dir.empty ()
+    ? subst_octave_home (std::string (OCTAVE_ARCHLIBDIR)) : dir;
+}
+
 // Adapted from libtool wrapper.
 #if defined (__WIN32__) && ! defined (__CYGWIN__)
 
 /* Prepares an argument vector before calling spawn().
    Note that spawn() does not by itself call the command interpreter
      (getenv ("COMSPEC") != NULL ? getenv ("COMSPEC") :
       ({ OSVERSIONINFO v; v.dwOSVersionInfoSize = sizeof(OSVERSIONINFO);
          GetVersionEx(&v);
@@ -588,38 +606,42 @@ int
 main (int argc, char **argv)
 {
   int retval = 0;
 
   bool start_gui = false;
   bool gui_libs = false;
 
   std::string octave_bindir = get_octave_bindir ();
+  std::string octave_archlibdir = get_octave_archlibdir ();
 
   std::string file
     = octave_bindir + dir_sep_char + "octave-cli-" OCTAVE_VERSION;;
 
   char **new_argv = new char * [argc + 1];
 
   int k = 1;
 
   bool warn_display = true;
 
   for (int i = 1; i < argc; i++)
     {
       if (! strcmp (argv[i], "--force-gui"))
         {
           start_gui = true;
           gui_libs = true;
-          file = octave_bindir + dir_sep_char;
-          #if defined (HAVE_OCTAVE_GUI)
-            file += "octave-gui-" OCTAVE_VERSION;
-          #else
-            file += "octave-cli-" OCTAVE_VERSION;
-          #endif
+#if defined (HAVE_OCTAVE_GUI)
+          // The Octave version number is already embedded in the
+          // octave_archlibdir directory name so we don't need to
+          // append it to the octave-gui file name.
+
+          file = octave_archlibdir + dir_sep_char + "octave-gui";
+#else
+          file = octave_bindir + dir_sep_char + "octave-cli-" OCTAVE_VERSION;
+#endif
           new_argv[k++] = argv[i];
         }
       else if (! strcmp (argv[i], "--no-gui-libs"))
         {
           // Run the version of Octave that is not linked with any GUI
           // libraries.  It may not be possible to do plotting or any
           // ui* calls, but it will be a little faster to start and
           // require less memory.  Don't pass the --no-gui-libs option
